# 0. Installing required Python Packages

In [283]:
!pip install torch==2.3.0 torchmetrics==1.4.0.post0 transformers==4.41.2 accelerate==0.30.1 datasets==2.19.1 tokenizers==0.19.1 evaluate==0.4.2 numpy==1.26.4 tqdm==4.66.4 tensorboard==2.16.2 matplotlib==3.9.0 


[notice] A new release of pip is available: 25.0.1 -> 25.1.1
[notice] To update, run: python.exe -m pip install --upgrade pip


# Project Finances: Risque default de credits

- **Context**:
    - 
    - 
    - 
- **Goal**:
    - 
    - 
    - 
- **Data**:
    - **Dataset #1**: [Kaggle Twitter US Airline Sentiment](https://www.kaggle.com/datasets/crowdflower/twitter-airline-sentiment)
- **Challenges**:
    - 
    - 
    - 

## Load and Explore
- Let's quick load and explore to see which one should be the best suitable for this project 

In [249]:
import pandas as pd

In [250]:
csv_file = 'Risque defaut de credit.csv'

In [251]:
import os

print("Répertoire courant :", os.getcwd())
print("Contenu :", os.listdir())


Répertoire courant : c:\Users\mamat\OneDrive\Bureau\python_session6
Contenu : ['.git', '.venv', 'Risque defaut de credit.csv', 'Risque_defaut_credit_traite.csv']


In [252]:
csv_file = 'Risque defaut de credit.csv' 

In [254]:
df = pd.read_csv(csv_file, engine='python', on_bad_lines='skip')
df.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501    1296599     5000.0       5000.0           4975.0   36 months   
1  1077430    1314167     2500.0       2500.0           2500.0   60 months   
2  1077175    1313524     2400.0       2400.0           2400.0   36 months   
3  1075269    1311441     5000.0       5000.0           5000.0   36 months   
4  1072053    1288686     3000.0       3000.0           3000.0   36 months   

   int_rate  installment grade sub_grade  ... total_bal_il il_util  \
0     10.65       162.87     B        B2  ...          NaN     NaN   
1     15.27        59.83     C        C4  ...          NaN     NaN   
2     15.96        84.33     C        C5  ...          NaN     NaN   
3      7.90       156.46     A        A4  ...          NaN     NaN   
4     18.64       109.43     E        E1  ...          NaN     NaN   

  open_rv_12m  open_rv_24m max_bal_bc all_util total_rev_hi_lim inq_fi  \
0         NaN          NaN        NaN      NaN              NaN    NaN   
1         NaN          NaN        NaN      NaN              NaN    NaN   
2         NaN          NaN        NaN      NaN              NaN    NaN   
3         NaN          NaN        NaN      NaN              NaN    NaN   
4         NaN          NaN        NaN      NaN              NaN    NaN   

  total_cu_tl  \
0         NaN   
1         NaN   
2         NaN   
3         NaN   
4         NaN   

  inq_last_12m;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;  
0  ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;...                     
1  ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;...                     
2  ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;...                     
3  ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;...                     
4  ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;...                     

[5 rows x 74 columns]

In [259]:
# print info of df
print('Dataset Info: ')
print(df.info())

Dataset Info: 
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 833400 entries, 0 to 833399
Data columns (total 74 columns):
 #   Column                                                                 Non-Null Count   Dtype  
---  ------                                                                 --------------   -----  
 0   id                                                                     833400 non-null  int64  
 1   member_id                                                              833400 non-null  int64  
 2   loan_amnt                                                              833400 non-null  float64
 3   funded_amnt                                                            833400 non-null  float64
 4   funded_amnt_inv                                                        833400 non-null  float64
 5   term                                                                   833400 non-null  object 
 6   int_rate                                                     

In [260]:
colonnes = df.columns.tolist()
print(colonnes)

['id', 'member_id', 'loan_amnt', 'funded_amnt', 'funded_amnt_inv', 'term', 'int_rate', 'installment', 'grade', 'sub_grade', 'emp_title', 'emp_length', 'home_ownership', 'annual_inc', 'verification_status', 'issue_d', 'loan_status', 'pymnt_plan', 'url', 'desc', 'purpose', 'title', 'zip_code', 'addr_state', 'dti', 'delinq_2yrs', 'earliest_cr_line', 'inq_last_6mths', 'mths_since_last_delinq', 'mths_since_last_record', 'open_acc', 'pub_rec', 'revol_bal', 'revol_util', 'total_acc', 'initial_list_status', 'out_prncp', 'out_prncp_inv', 'total_pymnt', 'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int', 'total_rec_late_fee', 'recoveries', 'collection_recovery_fee', 'last_pymnt_d', 'last_pymnt_amnt', 'next_pymnt_d', 'last_credit_pull_d', 'collections_12_mths_ex_med', 'mths_since_last_major_derog', 'policy_code', 'application_type', 'annual_inc_joint', 'dti_joint', 'verification_status_joint', 'acc_now_delinq', 'tot_coll_amt', 'tot_cur_bal', 'open_acc_6m', 'open_il_6m', 'open_il_12m', 'open_il

## Selecting main features for modeling sentiment classifier

In [262]:
df['loan_status'].value_counts(dropna=False)

loan_status
Current                                                586525
Fully Paid                                             176783
Charged Off                                             39465
Late (31-120 days)                                      11238
Issued                                                   8379
In Grace Period                                          6070
Late (16-30 days)                                        2294
Default                                                  1178
Does not meet the credit policy. Status:Fully Paid       1072
Does not meet the credit policy. Status:Charged Off       396
Name: count, dtype: int64

In [263]:
df1 = df[['loan_status']]
df1.head()

loan_status
0   Fully Paid
1  Charged Off
2   Fully Paid
3   Fully Paid
4   Fully Paid

In [264]:
# On définit la liste des colonnes/features
features = [
    'loan_amnt', 'term', 'int_rate', 'installment',
    'annual_inc', 'delinq_2yrs', 
    'emp_length', 'home_ownership', 'verification_status',
    'purpose'
]

## EDA

In [268]:
# print info of df
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 833400 entries, 0 to 833399
Data columns (total 1 columns):
 #   Column       Non-Null Count   Dtype 
---  ------       --------------   ----- 
 0   loan_status  833400 non-null  object
dtypes: object(1)
memory usage: 6.4+ MB


In [269]:
# Count the sentiment labels by class
df['loan_status'].value_counts().sort_values(ascending=False)

loan_status
Current                                                586525
Fully Paid                                             176783
Charged Off                                             39465
Late (31-120 days)                                      11238
Issued                                                   8379
In Grace Period                                          6070
Late (16-30 days)                                        2294
Default                                                  1178
Does not meet the credit policy. Status:Fully Paid       1072
Does not meet the credit policy. Status:Charged Off       396
Name: count, dtype: int64

## Clean data

In [ ]:
# 3. Nettoyage : suppression des colonnes inutiles
#df.drop(columns=["id", "member_id", "url", "desc", "title", "zip_code", "emp_title",
   # "tot_coll_amt", "inq_fi", "il_util", "last_credit_pull_d",
    #"policy_code", "next_pymnt_d", "collections_12_mths_ex_med",
    #"mths_since_last_major_derog",
    #"annual_inc_joint", "dti_joint", "verification_status_joint",
    #"acc_now_delinq",
    #"total_rec_prncp", "total_rec_int",
    #"recoveries", "issue_d", "initial_list_status",
    #"out_prncp", "out_prncp_inv"], inplace=True)

In [ ]:
# 4. Suppression des lignes icomplet
#df2 = df.loc[:, df.isnull().mean() < 0.7]

In [270]:
# 5. Création de la variable cible binaire : 1 = défaut, 0 = remboursé
defaut_mapping = {
    "Charged Off": 1,
    "Fully Paid": 0
}
df['target'] = df['loan_status'].map(defaut_mapping)
df = df[df['target'].notnull()]  # Ne garde que les lignes avec target définie

print(df['target'].value_counts(normalize=True))

target
0.0    0.817501
1.0    0.182499
Name: proportion, dtype: float64


In [ ]:
# 8. Suppression de la colonne d'origine
#df.drop(columns=['loan_status'], inplace=True,)

In [90]:
# 9. Remplissage des valeurs manquantes
#df.fillna(0, inplace=True)

C:\Users\mamat\AppData\Local\Temp\ipykernel_16180\2074183907.py:2: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df.fillna(0, inplace=True)


In [131]:
# On définit la liste des colonnes/features
#features = [
 #   'loan_amnt', 'term', 'int_rate', 'installment',
   # 'annual_inc', 'delinq_2yrs',  'emp_length', 'home_ownership', 'verification_status','purpose'
#]

In [271]:
# creation de la nouvelle base 

# 1.  On a déjà définit la liste des colonnes/features #features 

# 2.  on s'assure que toutes ces colonnes existent dans df
features_valides = [c for c in features if c in df.columns]
print(features_valides)


['loan_amnt', 'term', 'int_rate', 'installment', 'annual_inc', 'delinq_2yrs', 'emp_length', 'home_ownership', 'verification_status', 'purpose']


In [272]:

# 3. on crée un  nouveau DataFrame en ne gardant la target + features valides 
df_modele = df[['target'] + features_valides].copy()

# 4. Vérifiez
print("Dimensions du DataFrame modèle :", df_modele.shape)
df_modele.head()

Dimensions du DataFrame modèle : (216248, 11)


target  loan_amnt        term  int_rate  installment  annual_inc  \
0     0.0     5000.0   36 months     10.65       162.87     24000.0   
1     1.0     2500.0   60 months     15.27        59.83     30000.0   
2     0.0     2400.0   36 months     15.96        84.33     12252.0   
3     0.0     5000.0   36 months      7.90       156.46     36000.0   
4     0.0     3000.0   36 months     18.64       109.43     48000.0   

   delinq_2yrs emp_length home_ownership verification_status         purpose  
0          0.0  10+ years           RENT            Verified     credit_card  
1          0.0   < 1 year           RENT     Source Verified             car  
2          0.0  10+ years           RENT        Not Verified  small_business  
3          0.0    3 years           RENT     Source Verified         wedding  
4          0.0    9 years           RENT     Source Verified             car

## missing data

In [274]:

# on les compte
valeurs_manq = df_modele.isnull().sum()

# Pour voir les résultats
print("Nombre de valeurs manquantes par colonne :")
print(valeurs_manq)

Nombre de valeurs manquantes par colonne :
target                    0
loan_amnt                 0
term                      0
int_rate                  0
installment               0
annual_inc                0
delinq_2yrs               0
emp_length             9032
home_ownership            0
verification_status       0
purpose                   0
dtype: int64


##### traiter les valeurs manquants 

In [275]:
# j'ai fais le choix de les remplacer par 'Unknown'

df_modele['emp_length'] = df_modele['emp_length'].fillna('Unknown')
print(df_modele['emp_length'])

0         10+ years
1          < 1 year
2         10+ years
3           3 years
4           9 years
            ...    
833372    10+ years
833385     < 1 year
833387      2 years
833390    10+ years
833392      Unknown
Name: emp_length, Length: 216248, dtype: object


In [225]:
# 5. Suppression des lignes icomplet
#df_final = df_modele.loc[:, df.isnull().mean() < 0.7]

In [276]:

# 3) Afficher le nouveau comptage
print(missing_after['emp_length'])   # doit afficher 0
print("\nTous les na par colonne :")
print(missing_after)

0

Tous les na par colonne :
target                 0
loan_amnt              0
term                   0
int_rate               0
installment            0
annual_inc             0
delinq_2yrs            0
emp_length             0
home_ownership         0
verification_status    0
purpose                0
dtype: int64


In [277]:
# on compte le nombre d'obersvation dans notre data :
nombre_obs= df_modele.count()

# 2. Affichez le résultat :
print("Nombre d’observations (non-NA) par colonne :")
print(nombre_obs)

Nombre d’observations (non-NA) par colonne :
target                 216248
loan_amnt              216248
term                   216248
int_rate               216248
installment            216248
annual_inc             216248
delinq_2yrs            216248
emp_length             216248
home_ownership         216248
verification_status    216248
purpose                216248
dtype: int64


In [279]:
# 6. Vérifiez
print("Dimensions du DataFrame modèle :", df_modele.shape)
df_modele.head()

Dimensions du DataFrame modèle : (216248, 11)


target  loan_amnt        term  int_rate  installment  annual_inc  \
0     0.0     5000.0   36 months     10.65       162.87     24000.0   
1     1.0     2500.0   60 months     15.27        59.83     30000.0   
2     0.0     2400.0   36 months     15.96        84.33     12252.0   
3     0.0     5000.0   36 months      7.90       156.46     36000.0   
4     0.0     3000.0   36 months     18.64       109.43     48000.0   

   delinq_2yrs emp_length home_ownership verification_status         purpose  
0          0.0  10+ years           RENT            Verified     credit_card  
1          0.0   < 1 year           RENT     Source Verified             car  
2          0.0  10+ years           RENT        Not Verified  small_business  
3          0.0    3 years           RENT     Source Verified         wedding  
4          0.0    9 years           RENT     Source Verified             car

In [280]:
# 10. Sauvegarde du fichier nettoyé
df_modele.to_csv("Risque_defaut_credit_traite.csv", index=False)
print("Fichier nettoyé enregistré sous 'Risque_defaut_credit_traite.csv'.")

Fichier nettoyé enregistré sous 'Risque_defaut_credit_traite.csv'.


## Create balanced dataset for modeling
- *Remind that* the raw dataset has
    - fully paid:    176783 samples
    - charged off:     39465 samples
 - In this project, I choose downsampling approach to obtain the balanced dataset by choosing the cut-off sampling as the minimum class size (meaning 39465 of charged off class). This way, we still has reasonable number of sampling for modeling with machine learning and deep learning approaches.


In [281]:
# Separate the dataset by sentiment
defaults = df_final[df_final['target'] == 1]
paid     = df_final[df_final['target'] == 0]


# Determine the minimum class size

min_size = min(len(defaults), len(paid))
print(f"Taille avant équilibrage → défauts: {len(defaults)}, remboursés: {len(paid)}")
print("On équibre à", min_size, "observations par classe.")

# Balance the dataset by undersampling

defaults_balanced = defaults.sample(min_size, random_state=42)
paid_balanced     = paid.sample(min_size, random_state=42)



# Combine the balanced datasets
balanced_df = pd.concat([defaults_balanced, paid_balanced], axis=0)

# Shuffle the balanced dataset

# 5) Mélanger et remettre les index à plat
balanced_df = balanced_df.sample(frac=1, random_state=42).reset_index(drop=True)

# 6) Vérifier
print("\nAprès équilibrage → défauts:", 
      balanced_df['target'].value_counts().to_dict())

Taille avant équilibrage → défauts: 39465, remboursés: 176783
On équibre à 39465 observations par classe.

Après équilibrage → défauts: {1: 39465, 0: 39465}


- **Remarks**: Well, we really obtained a new balance dataset for the modeling step

## Split dataset into Train, Validation and Test subsets
- This step is important in modeling
    - For Training: we will use Train subset (and Validation subset if applicable)
    - For Testing: we will use Test subset
- We will use 'stratify=balanced_df['target']' to make sure that all the subsets are sampled equally regarding the defaults classes
- At the end, we will save all subsets into new csv file for later use

In [284]:
%pip install scikit-learn


Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 25.0.1 -> 25.1.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [285]:

%pip install scikit-learn --quiet



Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 25.0.1 -> 25.1.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [287]:
%pip install scipy scikit-learn

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 25.0.1 -> 25.1.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [242]:
# Split the dataset into train, validation, and test sets
train_df, temp_df = train_test_split(balanced_df, test_size=0.3, random_state=42, stratify=balanced_df['target'])
val_df, test_df = train_test_split(temp_df, test_size=0.5, random_state=42, stratify=temp_df['target'])

# Print the sizes of the datasets
print(f"Train set size: {len(train_df)}")
print(f"Validation set size: {len(val_df)}")
print(f"Test set size: {len(test_df)}")

# 5) Sauvegarde en CSV
train_df.to_csv('data/default_risk/train_default.csv',      index=False)
val_df.to_csv(  'data/default_risk/val_default.csv',        index=False)
test_df.to_csv( 'data/default_risk/test_default.csv',       index=False)

NameError: name 'train_test_split' is not defined